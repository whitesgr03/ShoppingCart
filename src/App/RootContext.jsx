import { createContext, useReducer, useContext } from "react";
import cartReducer from "../components/modals/modalCartReducer";
import modalReducer from "../components/modals/modalReducer";

const cartContext = createContext([]);
const cartDispatchContext = createContext(cartReducer);

const useCart = () => useContext(cartContext);
const useCartDispatch = () => useContext(cartDispatchContext);

const modalContext = createContext(null);
const modalDispatchContext = createContext(modalReducer);

const useModal = () => useContext(modalContext);
const useModalDispatch = () => useContext(modalDispatchContext);

const RootProvider = ({ children }) => {
	const [cart, cartListDispatch] = useReducer(cartReducer, []);
	const [modal, modalDispatch] = useReducer(modalReducer, {
		modal: "",
		alertProduct: {
			state: "",
			product: null,
		},
	});

	return (
		<modalContext.Provider value={modal}>
			<modalDispatchContext.Provider value={modalDispatch}>
				<cartContext.Provider value={cart}>
					<cartDispatchContext.Provider value={cartListDispatch}>
						{children}
					</cartDispatchContext.Provider>
				</cartContext.Provider>
			</modalDispatchContext.Provider>
		</modalContext.Provider>
	);
};

export {
	RootProvider as default,
	useCart,
	useCartDispatch,
	useModal,
	useModalDispatch,
};
