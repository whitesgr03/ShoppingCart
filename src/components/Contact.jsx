import "../style/contact.css";
import Icon from "@mdi/react";
import { mdiPhone, mdiEmailOutline, mdiMapMarker } from "@mdi/js";

import { useOutletContext } from "react-router-dom";

const Contact = () => {
	const { imageUrls } = useOutletContext();

	const url = imageUrls.find(imageUrl => imageUrl.includes("contact"));

	return (
		<div className="contact">
			<div
				style={{
					backgroundImage: url ? `url(${url})` : " none",
				}}
			>
				<div className="blur">
					<h2 className="title">Contact Us</h2>
					<fieldset className="container">
						<legend>Get in touch</legend>
						<div className="info">
							<ul className="icons">
								<li>
									<Icon path={mdiPhone} size={1} />
								</li>
								<li>
									<Icon path={mdiEmailOutline} size={1} />
								</li>
								<li>
									<Icon path={mdiMapMarker} size={1} />
								</li>
							</ul>
							<ul className="describe">
								<li>+123-456-789</li>
								<li>gentskin@example.com</li>
								<li>123 Anywhere St.,Any City</li>
							</ul>
						</div>
						<form id="contact" onSubmit={e => e.preventDefault()}>
							<label htmlFor="name">
								Name
								<input type="text" name="name" id="name" />
							</label>
							<label htmlFor="email">
								Email
								<input type="text" name="email" id="email" />
							</label>
							<label htmlFor="content">
								Content
								<textarea
									name="content"
									id="content"
									rows="5"
								></textarea>
							</label>
						</form>
					</fieldset>
					<div className="wrap">
						<button type="submit" form="contact" className="slide">
							Send Message
						</button>
					</div>
				</div>
			</div>
		</div>
	);
};

export { Contact as default };
