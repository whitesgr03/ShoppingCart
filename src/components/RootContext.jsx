import { createContext, useReducer, useContext } from "react";
import cartReducer from "./modals/modalCartReducer";
import modalReducer from "./modals/modalReducer";

const cartContext = createContext([]);
const cartDispatchContext = createContext(cartReducer);

const useCart = () => useContext(cartContext);
const useCartDispatch = () => useContext(cartDispatchContext);

const modalContext = createContext(null);
const modalDispatchContext = createContext(modalReducer);

const useModal = () => useContext(modalContext);
const useModalDispatch = () => useContext(modalDispatchContext);

const RootContext = ({ children }) => {
	const [cart, cartListDispatch] = useReducer(cartReducer, []);
	const [modal, modalDispatch] = useReducer(modalReducer, {
		modal: "",
		alertProduct: {
			state: "",
			product: null,
		},
	});

	return (
		<modalContext.Provider value={modal}>
			<modalDispatchContext.Provider value={modalDispatch}>
				<cartContext.Provider value={cart}>
					<cartDispatchContext.Provider value={cartListDispatch}>
						{children}
					</cartDispatchContext.Provider>
				</cartContext.Provider>
			</modalDispatchContext.Provider>
		</modalContext.Provider>
	);
};

export {
	RootContext as default,
	useCart,
	useCartDispatch,
	useModal,
	useModalDispatch,
};
