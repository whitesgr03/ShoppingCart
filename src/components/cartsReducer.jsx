const cartReducer = (cart, action) => {
	switch (action.type) {
		case "added": {
			return cart.find(item => item.productId === action.item.productId)
				? cart.map(item =>
						item.productId === action.item.productId
							? {
									...item,
									quantity:
										item.quantity + action.item.quantity,
							  }
							: item
				  )
				: [
						...cart,
						{
							...action.item,
							id: cart.length === 0 ? 0 : cart.at(-1).id + 1,
						},
				  ];
		}

		case "changed": {
			return cart.map(item =>
				item.id === action.item.id ? action.item : item
			);
		}

		case "deleted": {
			return cart.filter(item => item.id !== action.id);
		}

		default:
			throw Error("Unknown action: " + action.type);
	}
};

export default cartReducer;
