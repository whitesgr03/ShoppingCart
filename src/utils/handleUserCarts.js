import { initialFirestore, initialAuth } from "../firebase-config";

const getUserCart = async () => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart");

	try {
		const cartItems = await cartRef.get();
		return {
			message: "Get cart success.",
			success: true,
			data: cartItems.empty
				? []
				: cartItems.docs.map(item => ({ id: item.id, ...item.data() })),
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

const addUserCartItem = async item => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;
	const { id: productId, ...product } = item;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart");

	const productRef = cartRef.doc(productId);

	try {
		await productRef.set(product);
		return {
			message: "Add to cart success.",
			success: true,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

const updateUserCartItem = async item => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart")
		.doc(item.id);

	try {
		await cartRef.update(item);
		return {
			message: "Update cart item success.",
			success: true,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

const deleteUserCartItem = async productId => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart")
		.doc(productId);

	try {
		await cartRef.delete();
		return {
			message: "Delete cart item success.",
			success: true,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

export { getUserCart, addUserCartItem, updateUserCartItem, deleteUserCartItem };
