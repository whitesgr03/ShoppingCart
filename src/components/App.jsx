import "../style/app.css";

import { useState, useEffect } from "react";

import { Outlet } from "react-router-dom";

import Header from "./layout/Header";
import Footer from "./layout/Footer";
import Modal from "./modals/Modal";
import Error from "./Error";


import Loading from "./Loading";


import { initialAuth } from "../firebase-config";

import preLoadImage from "../utils/handlePreLoadImage";
import getStorageImage from "../utils/handleStorageImage";
import { getUserCart } from "../utils/handleUserCart";

import {
	handleCheckUser,
	handleRegisterUser,
} from "../utils/handleUserAccount";


const Root = () => {
	const [imageUrls, setImageUrls] = useState(null);
	const [error, setError] = useState(null);

	return (
		<RootContext>
			<div className="app">
				<Header />

				<div data-testid="content" className={"content"}>
					{loading && (
						<div data-testid="loading" className="loading">
							<Icon path={mdiLoading} spin={1} size={3} />
							Loading...
						</div>
					)}
					{error && <Error message={error} />}
					{!loading && !error && (
						<Outlet
							context={{
								imageUrls,
							}}
						/>
					)}
					<Footer />
				</div>
				<Modal />
			</div>
		</RootContext>
	);
};

export default Root;
