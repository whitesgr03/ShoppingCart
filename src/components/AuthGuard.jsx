import { useEffect, useState } from "react";
import { useOutletContext } from "react-router-dom";

import Loading from "./Loading";

import { handleGoogleLogin } from "../utils/handleUserAccount";

import PropTypes from "prop-types";

const AuthGuard = ({ children }) => {
	const [loading, setLoading] = useState(true);
	const { userId, onError } = useOutletContext();

	useEffect(() => {
		let ignore = false;
		try {
			!ignore && !userId && handleGoogleLogin();
			!ignore && userId && setLoading(false);
		} catch (error) {
			!ignore && onError(error);
		}
		return () => {
			ignore = true;
		};
	}, [userId, onError]);

	return <>{loading ? <Loading /> : children}</>;
};

AuthGuard.propTypes = {
	children: PropTypes.node.isRequired,
};

export { AuthGuard as default };
