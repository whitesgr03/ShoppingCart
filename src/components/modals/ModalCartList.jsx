import "../../style/modals/modalCartList.css";
import Icon from "@mdi/react";
import { mdiTrashCanOutline, mdiLoading } from "@mdi/js";

import { useState } from "react";

import { updateUserCartItem } from "../../utils/handleUserCart";

import PropTypes from "prop-types";

const ModalCartList = ({
	cart,
	userId,
	onError,
	onGetUserCart,
	onOpenModule,
}) => {
	const [loading, setLoading] = useState(false);

	const handleRemove = product => onOpenModule("alert", product, "remove");
	const handleChange = async product => {
		try {
			setLoading(true);
			await updateUserCartItem(product, userId);
			await onGetUserCart(userId);
		} catch (error) {
			console.error(error);
			onError("Service temporarily unavailable");
		} finally {
			setLoading(false);
		}
	};

	return (
		<div className="items">
			<div className="itemWarp">
				{cart.map(product => (
					<div className="item" key={product.id}>
						<div className="warp">
							<img src={product.url} alt={product.name} />
						</div>
						<div className="description">
							<div className="info">
								<div className="name">{product.name}</div>
								<div className="price">
									${product.price.toFixed(2)}
								</div>
							</div>

							<div className="subtotal">
								<label htmlFor={product.name}>
									Quantity
									<select
										data-testid="quantity"
										name="quantity"
										id={product.name}
										onChange={e =>
											handleChange({
												...product,
												quantity: +e.target.value,
											})
										}
										value={product.quantity}
									>
										{Array.from({
											length:
												product.quantity > 10
													? product.quantity
													: 10,
										})
											.fill([])
											.map((_, i) => (
												<option key={i} value={i + 1}>
													{i + 1}
												</option>
											))}
									</select>
								</label>
								<p>
									Subtotal: $
									{(product.quantity * product.price).toFixed(
										2
									)}
								</p>
							</div>
						</div>
						<button
							className="removeBtn"
							data-testid="removeBtn"
							onClick={() => handleRemove(product)}
						>
							<Icon path={mdiTrashCanOutline} />
						</button>
					</div>
				))}
			</div>

			{loading && (
				<div className="loading">
					<Icon path={mdiLoading} spin={1} size={3}></Icon>
				</div>
			)}
		</div>
	);
};

ModalCartList.propTypes = {
	list: PropTypes.array,
	isLoading: PropTypes.bool,
	onLoading: PropTypes.func,
};

ModalCartListOption.propTypes = {
	quantity: PropTypes.number,
};

export default ModalCartList;
