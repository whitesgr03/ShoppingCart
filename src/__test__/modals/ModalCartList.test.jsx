import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { RouterProvider, createMemoryRouter } from "react-router-dom";

import { updateUserCartItem } from "../../utils/handleUserCarts";
import { useModalDispatch, useCartDispatch } from "../../components/RootContext";

import ModalCartList from "../../components/modals/ModalCartList";

jest.mock("../../components/RootContext");
jest.mock("../../firebase-config");
jest.mock("../../utils/handleUserCarts");

describe("Renders ModalCartList Component", () => {
	it("Should return ModalCartList DOM", async () => {
		const mockCartData = [
			{
				id: 0,
				name: "fake",
				url: "../",
				price: 19.9,
				quantity: 1,
			},
		];

		const mockIsLoad = false;

		const mockOnLoading = jest.fn();

		const routes = [
			{
				path: "/",
				element: (
					<ModalCartList
						list={mockCartData}
						isLoading={mockIsLoad}
						onLoading={mockOnLoading}
					/>
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		const { container } = render(<RouterProvider router={router} />);

		expect(container).toMatchSnapshot();
	});
	it("Should show Remove Alert modal DOM with click event", async () => {
		const mockCartData = {
			id: 0,
			name: "fake",
			url: "../",
			price: 19.9,
			quantity: 1,
		};

		const fakeData = {
			type: "alert",
			item: {
				state: "remove",
				product: mockCartData,
			},
		};

		const mockIsLoad = false;

		const mockModalDispatch = jest.fn();
		const mockOnLoading = jest.fn();

		useModalDispatch.mockReturnValueOnce(mockModalDispatch);

		const user = userEvent.setup();

		const routes = [
			{
				path: "/",
				element: (
					<ModalCartList
						list={[mockCartData]}
						isLoading={mockIsLoad}
						onLoading={mockOnLoading}
					/>
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const button = screen.getByTestId("removeBtn");

		await user.pointer({ keys: "[MouseLeft]", target: button });

		expect(mockModalDispatch).toBeCalledTimes(1);
		expect(mockModalDispatch.mock.calls[0][0]).toEqual(fakeData);
	});
	it("Should change select value with change event", async () => {
		const mockCartData = {
			id: 0,
			name: "fake",
			url: "../",
			price: 19.9,
			quantity: 1,
		};

		const updateFetchResult = {
			message: "Update cart item success.",
			success: true,
		};

		const mockIsLoad = false;

		const mockModalDispatch = jest.fn();
		const mockCartDispatch = jest.fn();
		const mockOnLoading = jest.fn();

		useModalDispatch.mockReturnValueOnce(mockModalDispatch);
		useCartDispatch.mockReturnValueOnce(mockCartDispatch);
		updateUserCartItem.mockReturnValueOnce(updateFetchResult);

		const user = userEvent.setup();

		const routes = [
			{
				path: "/",
				element: (
					<ModalCartList
						list={[mockCartData]}
						isLoading={mockIsLoad}
						onLoading={mockOnLoading}
					/>
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const select = screen.getByTestId("quantity");

		expect(select).toHaveValue("1");

		await user.selectOptions(select, "2");

		expect(updateUserCartItem).toBeCalledTimes(1);
		expect(mockOnLoading).toBeCalledTimes(2);
		expect(mockCartDispatch).toBeCalledTimes(1);
	});
});
