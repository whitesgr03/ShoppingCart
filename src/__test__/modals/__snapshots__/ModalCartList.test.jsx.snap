// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ModalCartList Component Should return ModalCartList DOM 1`] = `
<div>
  <div
    class="items"
  >
    <div
      class="itemWarp"
    >
      <div
        class="item"
      >
        <div
          class="warp"
        >
          <img
            alt="fake"
            src="../"
          />
        </div>
        <div
          class="description"
        >
          <div
            class="info"
          >
            <div
              class="name"
            >
              fake
            </div>
            <div
              class="price"
            >
              $
              19.90
            </div>
          </div>
          <div
            class="subtotal"
          >
            <label
              for="fake"
            >
              Quantity
              <select
                data-testid="quantity"
                id="fake"
                name="quantity"
              >
                <option
                  value="1"
                >
                  1
                </option>
                <option
                  value="2"
                >
                  2
                </option>
                <option
                  value="3"
                >
                  3
                </option>
                <option
                  value="4"
                >
                  4
                </option>
                <option
                  value="5"
                >
                  5
                </option>
                <option
                  value="6"
                >
                  6
                </option>
                <option
                  value="7"
                >
                  7
                </option>
                <option
                  value="8"
                >
                  8
                </option>
                <option
                  value="9"
                >
                  9
                </option>
                <option
                  value="10"
                >
                  10
                </option>
              </select>
            </label>
            <p>
              Subtotal: $
              19.90
            </p>
          </div>
        </div>
        <button
          class="removeBtn"
          data-testid="removeBtn"
        >
          <svg
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z"
              style="fill: currentColor;"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
